docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,421 @@
+
+Generic Codecs
+--------------
+
+The previous example showed how to define a codec for a specific class. However, sometimes you want to define codecs to
+handle a given type in all documents. For such codecs, you can implement the ``MongoDB\Codec\Codec`` interface. The
+following example defines a codec to store ``DateTimeInterface`` instances as BSON dates including the timezone:
+
+.. code-block:: php
+
+   <?php
+
+   final class DateTimeCodec implements MongoDB\Codec\Codec
+   {
+       // These traits define commonly used functionality to avoid duplication
+       use MongoDB\Codec\DecodeIfSupported;
+       use MongoDB\Codec\EncodeIfSupported;
+
+       public function canDecode($value): bool
+       {
+           // For maximum compatibility, this codec supports decoding both UTCDateTime instances and documents with a
+           // UTCDateTime instance and optional timezone
+           return
+               $value instanceof MongoDB\BSON\UTCDateTime ||
+               $value instanceof MongoDB\BSON\Document && $value->has('utc');
+       }
+
+       public function canEncode($value): bool
+       {
+           return $value instanceof DateTimeInterface;
+       }
+
+       public function decode($value): DateTimeImmutable
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $utc = $value instanceof MongoDB\BSON\UTCDateTime
+               ? $value
+               : $value->get('utc');
+
+           if (! $utc instanceof UTCDateTime) {
+               throw UnsupportedValueException::invalidDecodableValue($utc);
+           }
+
+           $dateTime = $utc->toDateTime();
+
+           if ($value instanceof MongoDB\BSON\Document && $value->has('tz')) {
+               $dateTime->setTimeZone(new DateTimeZone($value->get('tz')));
+           }
+
+           return DateTimeImmutable::fromMutable($dateTime);
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'utc' => new UTCDateTime($value),
+               'tz' => $value->getTimezone()->getName(),
+           ]);
+       }
+   }
+
+This codec can now be leveraged by other codecs encode and decode dates. Let's return to the previous example and add a
+``createdAt`` field that contains the creation date. The modified encode and decode methods would look like this (other
+code omitted for brevity):
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       private DateTimeCodec $dateTimeCodec;
+
+       public function __construct()
+       {
+           $this->dateTimeCodec = new DateTimeCodec();
+       }
```

Comment:
For the doc, can we use PHP 8.1 features: promoted property and [new in initializer](https://wiki.php.net/rfc/new_in_initializers).
```suggestion
       public function __construct(
           private DateTimeCodec $dateTimeCodec = new DateTimeCodec(),
       ) {}
```